src/cargo/core/compiler/mod.rs

```diff
@@ -1347,13 +1369,21 @@ fn replay_output_cache(
     path: PathBuf,
     format: MessageFormat,
     color: bool,
+    show_warnings: bool,
 ) -> Work {
     let target = target.clone();
     let mut options = OutputOptions {
         format,
         look_for_metadata_directive: true,
         color,
         cache_cell: None,
+        // If we don't want to display warnings for this
+        // unit (e.g. we aren't in verbose mode, and a crate.io
+        // crate is being compiled), then we only want to replay
+        // future incompat report messages, not any other diagnostics
+        // (which would be suppressed by `--cap-lints` if we were
+        // doing a full rebuild).
+        replay_future_incompat_only: !show_warnings,
```

Comment:
Previously, we gated replaying messages on `unit.show_warnings`. We now need to replay messages for all units in order to detect future-incompat-report messages. However, we need to respect `show_warnings`, and avoid replaying warning messages for crates that we don't want to show warnings for.

Reply:
Do warnings get printed if this is removed? The `--cap-lints` flag should prevent the compiler from generating lints most of the time here I thought.

Reply:
The problem is when we have a sequence of commands like this:

1. `cargo build --verbose` (we don't pass `--cap-lints`, causing us to cache warning messages)
2. `cargo build` (we replay the cached messages, which includes warnings that we no longer want to show).

Previously, we would not replay any messages for the second `cargo build` invocation (due to `show_warnings` being `false`), so it didn't matter that we cached them.

Reply:
Hm ok I forgot that we tried to handle that. This is the support I was thinking of though where I would prefer to leave it to a separate subcommand if we can rather than editing lots of stuff in the main compile pipeline to do the right gymnastics with figuring out where warnings go.

Reply:
All of this logic is independent of the way in which we display the output (flag vs subcommand). If the user does:

1. `cargo build`
2. edit a file in the local crate
3. `cargo build`
4. Display the future incompat report (either through a command or flag)

Then the second `cargo build` invocation needs to replay cached messages from upstream crates in order to generate the correct report.

Actually, supporting *only* a flag made the implementation significantly less complicated, since  we no longer need to serialize anything to disk. Once we have the requirement of serializing this to disk (in order to support a command), the actual flag implementation is trivial - we just [need to check it before printing the report](https://github.com/Aaron1011/cargo/blob/3199de8b5de6f440cdb6e7a3c8401f351afa7344/src/cargo/core/compiler/job_queue.rs#L854-L863)